fix(appium): correctly apply extension defaults

boneskull opened the pull request:
Extensions were not correctly receiving the default values as specified in their schemas.  For example, in `xcuitest` driver, the `wdaLocalPort` option has a default value of `8100`, but this value would never make it to the driver.

The root cause was that `getDefaultsForSchema()` returns a _flattened_ object (which is useful for displaying to users as part of `getNonDefaultServerArgs()`), but by the time defaults should get assigned to parsed arguments in `main.js`, the parsed arguments object has already been expanded to a deeply-nested, non-flattened object.  So `"driver.driverName.argName": "value"` would be set in the args, but  not `{"driver": {"driverName": {"argName": "value"}}}`.

This change fixes the issue by iterating over the result of `getDefaultsForSchema()` and using `_.set()` to build a deeply-nested object from the defaults, which is then applied to the parsed arguments.

Closes #16165.

boneskull committed:
fix(appium): correctly apply extension defaults

Extensions were not correctly receiving the default values as specified in their schemas.  For example, in `xcuitest` driver, the `wdaLocalPort` option has a default value of `8100`, but this value would never make it to the driver.

The root cause was that `getDefaultsForSchema()` returns a _flattened_ object (which is useful for displaying to users as part of `getNonDefaultServerArgs()`), but by the time defaults should get assigned to parsed arguments in `main.js`, the parsed arguments object has already been expanded to a deeply-nested, non-flattened object.  So `"driver.driverName.argName": "value"` would be set in the args, but  not `{"driver": {"driverName": {"argName": "value"}}}`.  Furthermore, the properties of `cliArgs` are not prefixed with `<extType>.<extName>.`, but are just the arg name itself (which makes sense).

To address this, two main changes:

1. `getDefaultsForSchema(flatten = true)` (renamed from `getDefaultsFromSchema()`) now accepts a `flatten` flag, which will return a flat object using `<extType>.<extName>.`-style prefixes. This is useful (and less computationally intensive) for comparing defaults across the whole of Appium and its extensions.  This is _not_ useful when merging defaults into the parsed arguments object, however, as that object will be deeply-nested.  If `flatten` is `false`, a deeply-nested object will be returned, which is consumed by the `main` module when post-processing parsed CLI arguments and config file values.
2. Added `getDefaultsForExtension(extType, extName)` which acts like `getDefaultsForSchema(flatten = true)` _but_ only returns defaults for a particular extension _and_ strips the leading prefixes.  This is used by `assignCliArgsToExtension` in the `AppiumDriver` class.

Other changes:

1. `ArgSpec` now keeps a `rawDest` property which is like `dest` w/o the prefixes.
2. Make more types
3. Re-organize `driver-specs` unit test somewhat, and update to reflect changes.

boneskull commented on packages/appium/test/schema/schema-specs.js:
```diff
@@ -33,11 +33,6 @@ describe('schema', function () {
-      '../../lib/extension-config': {
-        DRIVER_TYPE: 'driver',
-        PLUGIN_TYPE: 'plugin',
-      },
```
this mock was just not needed

boneskull commented on packages/appium/test/parser-specs.js:
```diff
@@ -20,11 +20,11 @@ describe('parser', function () {
-      p.parse_args(['driver', 'list']);
-      (() => p.parse_args(['foo'])).should.throw();
-      (() => p.parse_args(['foo --bar'])).should.throw();
+      p.parseArgs([]);
```
`parseArgs()` is the canonical name of this method.  `parse_args()` is deprecated (and should probably just be removed). vestige from calling into `argparse` directly

mykola-mokhnach commented:
Seems like tests are quite unhappy

Also, I'm not quite sure about how default values are processed. The idea is that if an argument is not provided then it should not be passed to a driver at all, otherwise there might be an issue where a user passes a capability to a test session and then it still gets overridden by CLI arg, because it did accept a default value.

mykola-mokhnach commented on packages/appium/lib/main.js:
```diff
@@ -216,18 +216,18 @@ async function init (args) {
   if (parsedArgs.subcommand === SERVER_SUBCOMMAND) {
+    const flatDefaults = getDefaultsFromSchema();
+    const defaults = {};
+    for (const [key, value] of Object.entries(flatDefaults)) {
```
can we use Object.assign?

boneskull replied:
I'm not sure how.  Some keys in this object are of the format `extType.extName.argName` which would get ugly.

jlipps commented:
>  The idea is that if an argument is not provided then it should not be passed to a driver at all, otherwise there might be an issue where a user passes a capability to a test session and then it still gets overridden by CLI arg, because it did accept a default value.

Yeah this is an interesting point. It's nice to have driver args with defaults in a schema. But we want to make sure not to override capabilities if the args were the defaults. Not sure if there's any way for a driver to tell whether args were defaults? Is cliArgs populated with defaults or only if a user actually specifies something on the CLI?

boneskull commented:
> Yeah this is an interesting point. It's nice to have driver args with defaults in a schema. But we want to make sure not to override capabilities if the args were the defaults. Not sure if there's any way for a driver to tell whether args were defaults? Is cliArgs populated with defaults or only if a user actually specifies something on the CLI?

If we only assign to `cliArgs` if the user provided non-default values, then drivers would have to assign default values manually, which is not currently a thing that happens (and, imo, is undesirable behavior).

So I think what I could do is gather the default values and only overwrite the caps if the CLI args differ from those defaults.

I hope I'm making sense.

boneskull commented:
OK, I am confused.

>> The idea is that if an argument is not provided then it should not be passed to a driver at all, otherwise there might be an issue where a user passes a capability to a test session and then it still gets overridden by CLI arg, because it did accept a default value.

> Yeah this is an interesting point. It's nice to have driver args with defaults in a schema. But we want to make sure not to override capabilities if the args were the defaults. Not sure if there's any way for a driver to tell whether args were defaults? Is cliArgs populated with defaults or only if a user actually specifies something on the CLI?

It _sounds like_ if a driver wants to override caps with CLI args, this must be implemented in the driver itself.

Is this correct?  If so, I can provide an API to do so in this PR, and send another PR to `xcuitest` to use it.

boneskull commented:
(And: should the API be specific to the use-case, e.g., `overrideCapsWithCliArgs()` instead of building blocks for such a method?)

mykola-mokhnach commented:
> (And: should the API be specific to the use-case, e.g., overrideCapsWithCliArgs() instead of building blocks for such a method? I think it'd be useful to provide facilities for a driver to "query" its own defaults, fwiw)

I'd say we go with the current implementation and then I'll spend some time testing the actual implementation after the changes are published

boneskull commented:
@mykola-mokhnach We can't go with this implementation, because it's failing the build (for good reason).  

I'm not entirely sure how drivers are supposed to handle defaults right now (since `argsConstraints` did not have any notion of defaults).  But I'll make it so, at minimum, those default values do not wind up in `cliArgs`.

boneskull commented:
@jlipps @mykola-mokhnach This is now updated to reflect the requested behavior (and _should_ pass the build).

From the commit message:

> 
> Extensions were not correctly receiving the default values as specified in their schemas.  For example, in `xcuitest` driver, the `wdaLocalPort` option has a default value of `8100`, but this value would never make it to the driver.
> 
> The root cause was that `getDefaultsForSchema()` returns a _flattened_ object (which is useful for displaying to users as part of `getNonDefaultServerArgs()`), but by the time defaults should get assigned to parsed arguments in `main.js`, the parsed arguments object has already been expanded to a deeply-nested, non-flattened object.  So `"driver.driverName.argName": "value"` would be set in the args, but  not `{"driver": {"driverName": {"argName": "value"}}}`.  Furthermore, the properties of `cliArgs` are not prefixed with `<extType>.<extName>.`, but are just the arg name itself (which makes sense).
> 
> To address this, two main changes:
> 
> 1. `getDefaultsForSchema(flatten = true)` (renamed from `getDefaultsFromSchema()`) now accepts a `flatten` flag, which will return a flat object using `<extType>.<extName>.`-style prefixes. This is useful (and less computationally intensive) for comparing defaults across the whole of Appium and its extensions.  This is _not_ useful when merging defaults into the parsed arguments object, however, as that object will be deeply-nested.  If `flatten` is `false`, a deeply-nested object will be returned, which is consumed by the `main` module when post-processing parsed CLI arguments and config file values.
> 2. Added `getDefaultsForExtension(extType, extName)` which acts like `getDefaultsForSchema(flatten = true)` _but_ only returns defaults for a particular extension _and_ strips the leading prefixes.  This is used by `assignCliArgsToExtension` in the `AppiumDriver` class.
> 
> Other changes:
> 
> 1. `ArgSpec` now keeps a `rawDest` property which is like `dest` w/o the prefixes.
> 2. Make more types
> 3. Re-organize `driver-specs` unit test somewhat, and update to reflect changes.

boneskull commented:
To be clear: extensions will still need to apply default values manually--which I believe is a) the status quo and b) less-than-ideal--and I hope we can address that soon

boneskull commented on packages/appium/lib/appium.js:
```diff
@@ -118,14 +119,22 @@ class AppiumDriver extends BaseDriver {
+      const defaults = getDefaultsForExtension(extType, extName);
+      const cliArgs = _.isEmpty(defaults)
+        ? allCliArgsForExt
+        : _.omitBy(allCliArgsForExt, (value, key) => defaults[key] === value);
```
This will probably fail if the default is, say, an empty array.  I'll take care of that.

boneskull replied:
OK, this is addressed

jlipps commented:
OK great. Haven't reviewed the code yet but how you've described it is I think where we should land for this PR. I agree with you that it might make sense for there to be an appium-sponsored API for driver authors to deal with cli args and caps more ergonomically. In general, this is one problem with the fact that within appium and its driver ecosystem we have chosen to conflate everything into one object `this.opts`. It's convenient because drivers don't need to think about where parameters are coming from (CLI, appium itself, or users via caps). But it has all the problems you have now seen. Some other interface entirely (maybe a proxy object on basedriver that looks like `this.opts` but isn't actually a merged object, rather something that pulls from e.g. `this.cliArgs`, `this.driverOpts`, and `this.capabilities` with appropriate priority) might be a better idea, and maybe we want to tackle that for Appium 2.0. Not sure, and will be a big project since it may involve evaluating each use of `this.opts` throughout the official drivers.

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16181**  (this PR)
    * **PR #16194** 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

boneskull commented:
@jlipps I'll wait for you to review before merging.
